### ISSUE TITLE
further avoid steps for bike?

### ISSUE THREAD

**Issue by karussell** [graphhopper/graphhopper#2803]
See e.g. this [route](https://graphhopper.com/maps/?point=52.4726%2C-2.1645&point=52.480021%2C-2.166343&profile=bike) where we prefer the shorter path (969m) with over 70m steps (!) instead of the 1.3km [alternatives routes](https://graphhopper.com/maps/?point=52.4726%2C-2.1645&point=52.480021%2C-2.166343&profile=bike&layer=Omniscale&custom_model=%7B%22distance_influence%22%3A15%2C%22priority%22%3A%5B%7B%22if%22%3A%22road_class%3D%3DSTEPS%22%2C%22multiply_by%22%3A%220.1%22%7D%5D%7D). 

Not sure why we do not pick at least [this route](https://graphhopper.com/maps/?point=52.480021%2C-2.166343&point=52.476103%2C-2.165909&point=52.4726%2C-2.1645&profile=bike) it is also under 1km but contains only 11m steps.

Or maybe this is more a data problem as the path after the steps are marked as cycleway (!?)

Other real world examples:

 * https://graphhopper.com/maps/?point=54.105588%252C-2.790275&point=54.105676%252C-2.790023&profile=bike
 * https://graphhopper.com/maps/?point=54.119558%2C-2.777136&point=54.122602%2C-2.778895&profile=bike
 * https://graphhopper.com/maps/?point=55.896367%252C-3.539176&point=55.895712%252C-3.539402&profile=bike
 * https://graphhopper.com/maps/?point=51.018127%2C-3.103208&point=51.018692%2C-3.102821&profile=bike

**Comment by otbutz** [comment]
In determining the priority rating, we seem to ignore steps so that the lower average speed is the only incentive to choose a different route. As these sections tend to be small, a 2km/h speed limit might not be enough to force a different routing decision.

I just skimmed over the code but this part looks a bit sketchy:

https://github.com/graphhopper/graphhopper/blob/d77faa1b4d5298b2dbd8f4a380b3e89132fe3b2f/core/src/main/java/com/graphhopper/routing/util/parsers/BikeCommonAverageSpeedParser.java#L201-L204

So if `speed == PUSHING_SECTION_SPEED` we wouldn't lower the limit for `highway:steps` from 4km/h to 2km/h?

This comment referencing #634 also seems to be outdated?

https://github.com/graphhopper/graphhopper/blob/d77faa1b4d5298b2dbd8f4a380b3e89132fe3b2f/core/src/main/java/com/graphhopper/routing/util/parsers/BikeCommonAverageSpeedParser.java#L76-L79

Basically we still do this, but the logic for it is further down, which makes the comment and the pseudo-constant pretty useless.

**Comment by ratrun** [comment]
I'm pretty sure that the problem comes from the `bicycle=designated` tagging of the 70m [`steps`](https://www.openstreetmap.org/way/868661795). I added a test for it:
```
    @Test
    public void testSteps() {
        ReaderWay way = new ReaderWay(1);
        way.setTag("highway", "steps");
        assertPriorityAndSpeed(SLIGHT_AVOID, 2, way);

        way.setTag("bicycle", "designated");
        assertPriorityAndSpeed(VERY_NICE, 18, way);
    }
```
We need to apply a change such that the second test does not pass any longer. I'm going to provide a PR and also change the `CYCLEWAY_SPEED` code issue mentioned by @otbutz.

**Comment by mprins** [comment]
while you are looking at this; MTB seems to prefer going down the stairs... (that are too steep to ride)
- MTB: https://graphhopper.com/maps/?point=52.112429%2C5.072632&point=52.109906%2C5.078344&profile=mtb&layer=Omniscale
- bike: https://graphhopper.com/maps/?point=52.112429%2C5.072632&point=52.109906%2C5.078344&profile=bike&layer=Omniscale

**Comment by ratrun** [comment]
The PR fixes the originally discovered case and by chance also the MTB example from @mprins. I changed the speed for steps for the MTB profile because we should assume that steps are pushing sections even for MTBs. And why would somebody push the bicycle on a MTB faster compared to the other two bike profiles?

@karussell: I also had a looks on your other real world examples but do not quite understand the problems these should illustrate.

**Comment by ratrun** [comment]
I played around with the priority classification for `highway=steps` using the other real world examples above from @karussell. Even the worst classification of the steps with `REACH_DESTINATION` only solves example [3](https://graphhopper.com/maps/?point=55.896361%2C-3.539158_Folly+Burn+Path%2C+Livingston%2C+EH54+6RP%2C+Vereinigtes+Königreich&point=55.895723%2C-3.53941_Eliburn+Road%2C+Livingston%2C+EH54+6FG%2C+Vereinigtes+Königreich&profile=bike&layer=Omniscale) and [4](https://graphhopper.com/maps/?point=51.018127%2C-3.103208&point=51.018692%2C-3.102821&profile=bike&layer=Omniscale), but not [1](https://graphhopper.com/maps/?point=54.105626%2C-2.790067&point=54.105549%2C-2.790298&profile=bike&layer=Omniscale) and [2](https://graphhopper.com/maps/?point=54.119558%2C-2.777136&point=54.122602%2C-2.778895&profile=bike&layer=Omniscale). The problem with the example 3 and 4 is that the `highway=steps` segment is too short. As this will be the case with steps in most circumstances I think that we should implement the priorisation with assigning the `REACH_DESTINATION`. I'm going to update the PR accordingly.

**Comment by ratrun** [comment]
With the change for assigning `highway=steps` to priority `REACH_DESTINATION` our built in [Krems testcase ](http://localhost:8989/maps/?point=48.410987%2C15.599492&point=48.383419%2C15.659294&profile=bike&layer=OpenStreetMap) changes significantly from 
![ScreenshotCurrent](https://user-images.githubusercontent.com/6134544/234367230-d4e43ca9-a5dd-4605-890e-4805f74baec5.png)

into

![ScreenshotAfterChanges](https://user-images.githubusercontent.com/6134544/234367301-9354a9c6-b69e-432c-a95c-f818df874072.png)

I'm not sure if this significant increase in the distance from 6.9km to 9.3km is worth the avoidance of 17 meters steps. What do you think?

**Comment by ratrun** [comment]
After some fine tuning and running into further failed testcases in Monaco I consider the priority classification "BAD" as the best compromise for `highway=steps`. Please check the update in PR #2804.
